fedimint-api-client/src/query.rs

```diff
@@ -80,11 +85,28 @@ impl ErrorStrategy {
     pub fn process<R>(&mut self, peer: PeerId, error: PeerError) -> QueryStep<R> {
         assert!(self.errors.insert(peer, error).is_none());
 
-        if self.errors.len() == self.threshold {
+        let call_error_count = self.errors.iter().filter(|(_, e)| e.is_call_err()).count();
+        let non_call_error_count = self.errors.len() - call_error_count;
+        if non_call_error_count == self.num_peers.one_honest() {
+            // If there enough non-application errors that there's no way to get consensus,
+            // there's no reason to continue.
+            QueryStep::Failure {
+                general: Some(anyhow!(
+                    "Received {} out of {} non-call errors from peers: {}",
+                    self.num_peers.threshold(),
+                    self.num_peers,
+                    self.format_errors()
+                )),
+                peers: mem::take(&mut self.errors),
+            }
+        } else if call_error_count == self.num_peers.threshold() {
+            // For a call-errors to surface as a federation-level, it needs get a threshold
+            // of responses being call-errors.
             QueryStep::Failure {
                 general: Some(anyhow!(
-                    "Received errors from {} peers: {}",
-                    self.threshold,
+                    "Received {} out of {} call errors from peers: {}",
+                    self.num_peers.threshold(),
+                    self.num_peers,
                     self.format_errors()
```

Comment:
We need the different threshold here since a query strategy should fail on f + 1 rpc errors, however on call errors (like TransactionError) we need a threshold of 2f + 1 to achieve consensus; if we use f+1 for call errors. Imo rpc errors and TransactionErrors are two very different concepts and should not be squashed into the same struct just because they are both called "Errors".

Reply:
Squashing these errors requires us to always keep in mind to call is_call_err to differentiate when handling them in the client; for example only retry when is_call_err returns false.